### DISCUSSION TITLE
Counterfactuals make inference customization difficult

### DISCUSSION THREAD

**Discussion by eb8680** [BasisResearch/chirho#127] [Ideas]
Models expanded by `MultiWorldCounterfactual` introduce plates automatically in a way that's hard to anticipate, and may mangle names of sample sites, e.g. by reusing `obs_mask` from `pyro.sample` to designate which copies of a particular variable are observed or unobserved (see #55 ). This makes it difficult to write custom guides or other custom inference components for estimating causal effects.

Most of the release examples (including the CEVAE #11) currently use autoguides or exact inference to get around this. Maybe that's just the recommendation we should be giving, including by using fancier autoguides like `AutoStructured` in more advanced examples? If so, we should discuss this explicitly somewhere in the documentation.

Alternatively, maybe there's a generic way to transform a guide that's correct for an unexpanded model into one that's correct for the expanded version, or documentation we could provide about customizing inference in this setting?

**Comment by SamWitty** [comment]
Alternatively, we could insist that any custom inference only operate on latents in the original (untransformed) model. This would impose some restrictions (e.g. we wouldn't be able to meaningfully condition on any random variables that are descendants of counterfactual outcomes, which could be useful for a Bayesian version of sensitivity analysis https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3800481/), but may be worth it.

To accomplish this we would want to;
1. automatically mask any new latents generated from the counterfactual handlers to silence any "missing variables in the guide" warnings and ...
2. leave names of factual variables unchanged

**Comment by robertness** [comment]
> we wouldn't be able to meaningfully condition on any random variables that are descendants of counterfactual outcomes

This seems like it would be required to have multiworld counterfactuals as a valid use case
